File: nix.go
package main

import (
	"bufio"
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"os"
	"os/exec"
	"sort"
	"strconv"
	"strings"
	"time"
)

const warmProjection = `
let
  flake = builtins.getFlake %s;
  requested = builtins.fromJSON %s;
  enabled = builtins.fromJSON %s;
  currentSystem = %s;
  sources = {
    nixos = flake.nixosConfigurations or {};
    darwin = flake.darwinConfigurations or {};
    home = flake.homeConfigurations or {};
  };
  names = kind:
    let configs = sources.${kind};
    in if requested == []
      then builtins.attrNames configs
      else builtins.filter (name: builtins.hasAttr name configs) requested;
  project = kind: name:
    let
      configs = sources.${kind};
      config = configs.${name};
      build =
        if kind == "home" then config.activationPackage
        else if kind == "darwin" then config.system
        else config.config.system.build.toplevel;
      filterSystem =
        if currentSystem == ""
        then build.system
        else config.pkgs.stdenv.buildPlatform.system or build.system;
    in
      if currentSystem != "" && filterSystem != currentSystem
      then filterSystem
      else builtins.deepSeq {
        system = build.system;
        path = build.outPath;
        drv = build.drvPath;
      } true;
in builtins.deepSeq (map (kind: map (project kind) (names kind)) enabled) true
`

const discoveryProjection = `
let
  flake = builtins.getFlake %s;
  enabled = builtins.fromJSON %s;
  sources = {
    nixos = flake.nixosConfigurations or {};
    darwin = flake.darwinConfigurations or {};
    home = flake.homeConfigurations or {};
  };
  present = {
    nixos = flake ? nixosConfigurations;
    darwin = flake ? darwinConfigurations;
    home = flake ? homeConfigurations;
  };
  forEnabled = value: builtins.listToAttrs (map (kind: {
    name = kind;
    value = value kind;
  }) enabled);
in {
  available = forEnabled (kind: builtins.attrNames sources.${kind});
  present = builtins.any (kind: present.${kind}) enabled;
}
`

const configurationProjection = `
let
  flake = builtins.getFlake %s;
  kind = %s;
  name = %s;
  currentSystem = %s;
  configs =
    if kind == "home" then flake.homeConfigurations
    else if kind == "darwin" then flake.darwinConfigurations
    else flake.nixosConfigurations;
  config = configs.${name};
  build =
    if kind == "home" then config.activationPackage
    else if kind == "darwin" then config.system
    else config.config.system.build.toplevel;
  filterSystem =
    if currentSystem == ""
    then build.system
    else config.pkgs.stdenv.buildPlatform.system or build.system;
in
  if currentSystem != "" && filterSystem != currentSystem
  then { system = filterSystem; path = ""; drv = ""; skipped = true; }
  else
    let result = {
      system = build.system;
      path = build.outPath;
      drv = build.drvPath;
      skipped = false;
    };
    in builtins.deepSeq result result
`

type configuration struct {
	System  string `json:"system"`
	Path    string `json:"path"`
	Drv     string `json:"drv"`
	Skipped bool   `json:"skipped"`
}

type configurationSet map[string]map[string]configuration

type evaluation struct {
	Configurations configurationSet
	Available      groupedNames
	Present        bool
	EvalTimes      map[string]time.Duration
}

type discovery struct {
	Available groupedNames `json:"available"`
	Present   bool         `json:"present"`
}

type closure struct {
	Size  int64
	Paths int
}

func warm(opts options, enabled []configurationKind, system string) error {
	flake, err := flakeURL(opts.flake)
	if err != nil {
		return err
	}
	requested, _ := json.Marshal(opts.names)
	keys := make([]string, len(enabled))
	for i, kind := range enabled {
		keys[i] = kind.Key
	}
	kinds, _ := json.Marshal(keys)
	var result bool
	return nixEval(fmt.Sprintf(
		warmProjection,
		nixString(flake),
		nixString(string(requested)),
		nixString(string(kinds)),
		nixString(system),
	), &result)
}

func evaluate(
	opts options,
	enabled []configurationKind,
	system string,
	progress *liveReport,
) (evaluation, error) {
	flake, err := flakeURL(opts.flake)
	if err != nil {
		return evaluation{}, err
	}
	keys := make([]string, len(enabled))
	for i, kind := range enabled {
		keys[i] = kind.Key
	}
	kinds, _ := json.Marshal(keys)
	var found discovery
	if err := nixEval(fmt.Sprintf(
		discoveryProjection,
		nixString(flake),
		nixString(string(kinds)),
	), &found); err != nil {
		return evaluation{}, err
	}

	result := evaluation{
		Configurations: make(configurationSet, len(enabled)),
		Available:      found.Available,
		Present:        found.Present,
		EvalTimes:      make(map[string]time.Duration),
	}
	selected := make(groupedNames, len(enabled))
	for _, kind := range enabled {
		result.Configurations[kind.Key] = make(map[string]configuration)
		selected[kind.Key] = selectNames(opts.names, result.Available[kind.Key])
		if progress != nil {
			for _, name := range selected[kind.Key] {
				progress.discover(kind.Key, name)
			}
		}
	}
	if progress != nil {
		progress.ready()
	}

	for _, kind := range enabled {
		for _, name := range selected[kind.Key] {
			expression := fmt.Sprintf(
				configurationProjection,
				nixString(flake),
				nixString(kind.Key),
				nixString(name),
				nixString(system),
			)
			if progress != nil {
				progress.start(kind.Key, name)
			}
			started := time.Now()
			var config configuration
			err := nixEval(expression, &config)
			duration := time.Since(started)
			if err != nil {
				if progress != nil {
					progress.abort()
				}
				return evaluation{}, fmt.Errorf("evaluate %s configuration %q: %w", kind.Label, name, err)
			}
			result.Configurations[kind.Key][name] = config
			if config.Skipped {
				if progress != nil {
					progress.skipped(kind.Key, name, config.System)
				}
				continue
			}
			result.EvalTimes[evaluationID(kind.Key, name)] = duration
			if progress != nil {
				progress.done(kind.Key, name, config.System, duration)
			}
		}
	}
	return result, nil
}

func selectNames(requested, available []string) []string {
	if len(requested) == 0 {
		return available
	}
	found := make(map[string]struct{}, len(available))
	for _, name := range available {
		found[name] = struct{}{}
	}
	selected := make([]string, 0, len(requested))
	for _, name := range requested {
		if _, ok := found[name]; ok {
			selected = append(selected, name)
			delete(found, name)
		}
	}
	return selected
}

func flakeURL(flake string) (string, error) {
	output, err := nixOutput(
		[]string{"flake", "metadata", "--quiet", "--json", flake},
	)
	if err != nil {
		return "", err
	}
	var metadata struct {
		URL string `json:"url"`
	}
	if err := json.Unmarshal(output, &metadata); err != nil {
		return "", fmt.Errorf("decode flake metadata: %w", err)
	}
	if metadata.URL == "" {
		return "", fmt.Errorf("nix reported an empty flake URL")
	}
	return metadata.URL, nil
}

func realise(kinds []configurationKind, selected groupedNames, result evaluation, progress *liveReport) error {
	derivations := make(map[string]struct{})
	for _, kind := range kinds {
		for _, name := range selected[kind.Key] {
			id := evaluationID(kind.Key, name)
			if _, ok := result.EvalTimes[id]; !ok {
				return fmt.Errorf(
					"missing evaluation timing for %s configuration %q",
					kind.Label,
					name,
				)
			}
			derivations[result.Configurations[kind.Key][name].Drv] = struct{}{}
		}
	}
	args := []string{"--realise", "--quiet", "--no-build-output"}
	for drv := range derivations {
		args = append(args, drv)
	}
	sort.Strings(args[3:])
	cmd := exec.Command("nix-store", args...)
	var diagnostics bytes.Buffer
	cmd.Stdin = os.Stdin
	cmd.Stdout = io.Discard
	cmd.Stderr = &diagnostics
	if err := cmd.Run(); err != nil {
		progress.abort()
		_, _ = diagnostics.WriteTo(os.Stderr)
		return fmt.Errorf("realise selected configuration closures: %w", err)
	}
	return nil
}

func closureStats(path, kind, name string, progress *liveReport) (closure, error) {
	progress.beginClosure(kind, name)
	defer progress.endClosure()

	cmd := exec.Command("nix", "path-info", "--recursive", "--size", path)
	cmd.Env = append(os.Environ(), "CLICOLOR_FORCE=0", "NO_COLOR=1")
	cmd.Stdin = os.Stdin
	stdout, err := cmd.StdoutPipe()
	if err != nil {
		return closure{}, fmt.Errorf("capture nix path-info output: %w", err)
	}
	var diagnostics bytes.Buffer
	cmd.Stderr = &diagnostics
	if err := cmd.Start(); err != nil {
		return closure{}, fmt.Errorf("start nix path-info: %w", err)
	}

	var stats closure
	scanner := bufio.NewScanner(stdout)
	scanner.Buffer(make([]byte, 4096), 16<<20)
	var parseErr error
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		separator := strings.LastIndexAny(line, " \t")
		if separator < 1 {
			if parseErr == nil {
				parseErr = fmt.Errorf("invalid nix path-info line %q", line)
			}
			continue
		}
		pathSize, err := strconv.ParseInt(line[separator+1:], 10, 64)
		if err != nil {
			if parseErr == nil {
				parseErr = fmt.Errorf("invalid nix path-info size: %w", err)
			}
			continue
		}
		stats.Size += pathSize
		stats.Paths++
		progress.closure(kind, name, stats)
	}
	scanErr := scanner.Err()
	runErr := cmd.Wait()
	if diagnostics.Len() > 0 {
		progress.abort()
		_, _ = diagnostics.WriteTo(os.Stderr)
	}
	if scanErr != nil {
		return closure{}, fmt.Errorf("read nix path-info output: %w", scanErr)
	}
	if parseErr != nil {
		return closure{}, parseErr
	}
	if runErr != nil {
		return closure{}, fmt.Errorf("nix path-info failed: %w", runErr)
	}
	return stats, nil
}

func nixSystem() (string, error) {
	output, err := nixOutput([]string{"config", "show", "system"})
	if err != nil {
		return "", err
	}
	system := strings.TrimSpace(string(output))
	if system == "" {
		return "", fmt.Errorf("nix reported an empty system")
	}
	return system, nil
}

func nixEval(expression string, value any) error {
	output, err := nixOutput(
		[]string{"eval", "--impure", "--quiet", "--json", "--expr", expression},
	)
	if err != nil {
		return err
	}
	if err := json.Unmarshal(output, value); err != nil {
		return fmt.Errorf("decode nix output: %w", err)
	}
	return nil
}

func nixOutput(args []string) ([]byte, error) {
	cmd := exec.Command("nix", args...)
	cmd.Env = append(os.Environ(), "CLICOLOR_FORCE=0", "NO_COLOR=1")
	operation := args[0]
	if operation == "flake" && len(args) > 1 {
		operation += " " + args[1]
	}
	var output bytes.Buffer
	cmd.Stdout = &output
	cmd.Stderr = os.Stderr
	if err := cmd.Run(); err != nil {
		return nil, fmt.Errorf("nix %s failed: %w", operation, err)
	}
	return output.Bytes(), nil
}

func evaluationID(kind, name string) string {
	return kind + "\x00" + name
}

func nixString(value string) string {
	return `"` + strings.NewReplacer(
		`\`, `\\`, `"`, `\"`, `${`, `\${`,
	).Replace(value) + `"`
}

File: src/nix.rs
use anyhow::{Context, Result, bail};
use serde::{Deserialize, de::DeserializeOwned};
use serde_json::json;
use std::{
    collections::{BTreeMap, BTreeSet, VecDeque},
    process::{Command, Stdio},
    sync::{Mutex, mpsc},
    thread,
    time::{Duration, Instant},
};

use crate::{Kind, args::Options, report::Live};

#[derive(Deserialize)]
pub struct Discovery {
    pub available: BTreeMap<String, Vec<String>>,
    pub present: bool,
}

#[derive(Deserialize)]
pub struct Configuration {
    pub system: String,
    pub drv: String,
    pub skipped: bool,
}

pub struct Evaluated {
    pub kind: Kind,
    pub name: String,
    pub config: Configuration,
    pub duration: Duration,
}

#[derive(Clone, Copy)]
pub struct Closure {
    pub size: u64,
    pub paths: usize,
}

fn output(args: &[&str], progress: &Live, label: &str) -> Result<Vec<u8>> {
    let mut child = Command::new("nix")
        .env("CLICOLOR_FORCE", "0")
        .env("NO_COLOR", "1")
        .stdin(Stdio::inherit())
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .args(args)
        .spawn()
        .context("start nix")?;
    let stderr = child.stderr.take().unwrap();
    let output = thread::scope(|scope| -> Result<_> {
        let diagnostics = scope.spawn(|| progress.diagnostics(stderr, label));
        let output = child.wait_with_output().context("wait for nix");
        diagnostics
            .join()
            .unwrap()
            .context("read nix diagnostics")?;
        output
    })?;
    if !output.status.success() {
        bail!("nix {} failed: {}", args[0], output.status);
    }
    Ok(output.stdout)
}

pub fn system(progress: &Live) -> Result<String> {
    let bytes = output(&["config", "show", "system"], progress, "")?;
    let system = String::from_utf8(bytes)?.trim().to_owned();
    if system.is_empty() {
        bail!("nix reported an empty system");
    }
    Ok(system)
}

pub fn flake_url(flake: &str, progress: &Live) -> Result<String> {
    #[derive(Deserialize)]
    struct Metadata {
        url: String,
    }
    let metadata: Metadata = serde_json::from_slice(&output(
        &[
            "flake",
            "metadata",
            "--no-update-lock-file",
            "--quiet",
            "--json",
            flake,
        ],
        progress,
        "",
    )?)
    .context("decode flake metadata")?;
    if metadata.url.is_empty() {
        bail!("nix reported an empty flake URL");
    }
    Ok(metadata.url)
}

fn nix_string(value: &str) -> String {
    format!(
        "\"{}\"",
        value
            .replace('\\', "\\\\")
            .replace('"', "\\\"")
            .replace("${", "\\${")
    )
}

pub struct Evaluator<'a> {
    pub opts: &'a Options,
    pub kinds: &'a [Kind],
    pub flake: &'a str,
    pub system: &'a str,
    pub progress: &'a Live,
}

impl Evaluator<'_> {
    pub fn eval<T: DeserializeOwned>(&self, mode: &str, kind: &str, name: &str) -> Result<T> {
        let arguments = json!({
            "flake": self.flake, "enabled": self.kinds.iter().map(|kind| kind.output).collect::<Vec<_>>(),
            "currentSystem": self.system,
            "mode": mode, "kind": kind, "name": name,
        });
        let expression = format!(
            "({}) (builtins.fromJSON {})",
            include_str!("projection.nix"),
            nix_string(&arguments.to_string())
        );
        let label = if mode == "configuration" {
            format!("{kind}.{name}")
        } else {
            String::new()
        };
        serde_json::from_slice(&output(
            &[
                "eval",
                "--impure",
                "--quiet",
                "--json",
                "--expr",
                &expression,
            ],
            self.progress,
            &label,
        )?)
        .context("decode nix evaluation")
    }

    pub fn evaluate(&self, available: &BTreeMap<String, Vec<String>>) -> Result<Vec<Evaluated>> {
        let mut pending = VecDeque::new();
        for &kind in self.kinds {
            let names = &available[kind.output];
            let mut seen = BTreeSet::new();
            let selected = if self.opts.names.is_empty() {
                names
            } else {
                &self.opts.names
            };
            for name in selected {
                if !self.opts.names.is_empty() && (!names.contains(name) || !seen.insert(name)) {
                    continue;
                }
                pending.push_back((kind, name));
            }
        }
        let total = pending.len();
        let pending = Mutex::new(pending);
        thread::scope(|scope| {
            let (sender, receiver) = mpsc::channel();
            for _ in 0..self.opts.jobs.min(total) {
                let sender = sender.clone();
                let pending = &pending;
                scope.spawn(move || {
                    loop {
                        let Some((kind, name)) = pending.lock().unwrap().pop_front() else {
                            break;
                        };
                        let started = Instant::now();
                        let result = self
                            .eval("configuration", kind.output, name)
                            .with_context(|| {
                                format!("evaluate {} configuration {name:?}", kind.label)
                            })
                            .map(|config| Evaluated {
                                kind,
                                name: name.clone(),
                                config,
                                duration: started.elapsed(),
                            });
                        if result.is_err() {
                            pending.lock().unwrap().clear();
                        }
                        if sender.send(result).is_err() {
                            break;
                        }
                    }
                });
            }
            drop(sender);
            let mut result = Vec::with_capacity(total);
            for row in receiver {
                result.push(row?);
            }
            Ok(result)
        })
    }
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct Build {
    drv_path: String,
    outputs: BTreeMap<String, String>,
}

pub fn realise(configs: &[Evaluated], progress: &Live) -> Result<BTreeMap<String, String>> {
    let drvs: BTreeSet<_> = configs
        .iter()
        .filter(|row| !row.config.skipped)
        .map(|row| row.config.drv.as_str())
        .collect();
    let installables: Vec<_> = drvs.iter().map(|drv| format!("{drv}^out")).collect();
    let mut args = vec!["build", "--no-link", "--json", "--quiet"];
    args.extend(installables.iter().map(String::as_str));
    let builds: Vec<Build> = serde_json::from_slice(
        &output(&args, progress, "").context("realise selected configuration closures")?,
    )
    .context("decode realized outputs")?;
    let mut paths = BTreeMap::new();
    for build in builds {
        let path = build
            .outputs
            .get("out")
            .context("realized configuration has no out output")?;
        paths.insert(build.drv_path, path.clone());
    }
    for drv in drvs {
        if !paths.contains_key(drv) {
            bail!("nix did not return an output for {drv}");
        }
    }
    Ok(paths)
}

pub fn closure(path: &str, progress: &Live) -> Result<Closure> {
    // Nix and Lix use different JSON schemas. Their path/size columns share
    // the same format, and store path names cannot contain whitespace.
    let bytes = output(&["path-info", "--recursive", "--size", path], progress, "")?;
    let text = std::str::from_utf8(&bytes).context("decode closure paths")?;
    let mut closure = Closure { size: 0, paths: 0 };
    for line in text.lines() {
        let mut columns = line.split_whitespace();
        columns.next().context("missing closure path")?;
        let nar_size: u64 = columns
            .next()
            .context("missing NAR size")?
            .parse()
            .context("invalid NAR size")?;
        if columns.next().is_some() {
            bail!("invalid nix path-info line {line:?}");
        }
        closure.size = closure
            .size
            .checked_add(nar_size)
            .context("closure size overflow")?;
        closure.paths += 1;
    }
    Ok(closure)
}
